src/map/storage.c

```diff
@@ -212,21 +220,28 @@ int storage_additem(struct map_session_data* sd, struct item* item_data, int amo
  *------------------------------------------*/
 int storage_delitem(struct map_session_data* sd, int n, int amount)
 {
+	struct item *it = NULL;
+
 	nullpo_retr(1, sd);
-	Assert_retr(1, n >= 0 && n < MAX_STORAGE);
-	if( sd->status.storage.items[n].nameid == 0 || sd->status.storage.items[n].amount < amount )
+	Assert_retr(1, n >= 0 && n < VECTOR_LENGTH(sd->storage));
+
+	it = &VECTOR_INDEX(sd->storage, n);
+
+	if (it->nameid == 0 || it->amount < amount)
 		return 1;
 
-	sd->status.storage.items[n].amount -= amount;
-	if( sd->status.storage.items[n].amount == 0 )
-	{
-		memset(&sd->status.storage.items[n],0,sizeof(sd->status.storage.items[0]));
-		sd->status.storage.storage_amount--;
-		if( sd->state.storage_flag == STORAGE_FLAG_NORMAL )
-			clif->updatestorageamount(sd, sd->status.storage.storage_amount, MAX_STORAGE);
+	it->amount -= amount;
+
+	if (it->amount == 0) {
+		VECTOR_ERASE(sd->storage, n);
```

Comment:
This will cause the storage to be re-sorted, which the client doesn't expect. Any data in an open storage window will likely become invalid, and the client will start using wrong indexes for subsequent operations. This needs to be tested, but I don't expect it to work properly (I'm not sure how to let the client know that we changed the indexes from what we originally sent)